### ISSUE TITLE
如何分数据集

### ISSUE THREAD

**Issue by JunmeiHao** [graytowne/caser_pytorch#8]
Following[17,33,35],we hold the first70% of actions in each user’s sequence as the training set and use the next 10%of actions as the validation set to search the optimal hyperparameter settings for all models.There maining 20%actions in each user’s sequence are used as the test set for evaluating a model’s performance

你好，我现在在分数据集，您可以把数据集的部分代码公开一下嘛   或者，你可以提示一下怎么按照用户序列分数据集，谢谢~非常感激！

**Comment by graytowne** [comment]
Hi @JunmeiHao, 感谢你对我们paper的关注，目前数据划分这边的代码是和之前的project整合在一起的没有单独拿出来，写的也比较乱所以就没有选择make public。如果你在数据集划分方面遇到什么问题的话，你可以把代码的链接发给我我可以帮你参考一下。

**Comment by graytowne** [comment]
I tentatively close this issue for now, if you have any further problems, just leave any message in this thread and I will reply you whenever I see it.

**Comment by JunmeiHao** [comment]
非常感谢您的回复，那我再仔细想一下，如果想不到我可能会再麻烦您。先谢谢您啦。您的论文真的很棒。我会把您的论文引用在我的论文中的。谢谢您。

在 2019-05-01 05:48:54，"Jiaxi Tang" <notifications@github.com> 写道：


Hi @JunmeiHao, 感谢你对我们paper的关注，目前数据划分这边的代码是和之前的project整合在一起的没有单独拿出来，写的也比较乱所以就没有选择make public。如果你在数据集划分方面遇到什么问题的话，你可以把代码的链接发给我我可以帮你参考一下。

**Comment by JunmeiHao** [comment]
您好，我想问一下，您可以把除了movielens数据集，其他三个分好的数据集发一下嘛，谢谢您

Jiaxi Tang <notifications@github.com>编写：

**Comment by 01joy** [comment]
您好，阅读了您的paper，很有启发性的idea，很棒。

关于推荐系统的数据集划分，paper中是这么描述的：

> Following[17,33,35],we hold the first70% of actions in each user’s sequence as the training set and use the next 10%of actions as the validation set to search the optimal hyperparameter settings for all models.There maining 20%actions in each user’s sequence are used as the test set for evaluating a model’s performance

也就是说train, validation, test的划分，是针对每个user的items进行划分的。有个问题是，在test的时候，items集合应该是全集吧，也就是说包含train, validation, test三部分中出现的items全集，那在validation和test的时候，模型肯定会输出train的时候的item吧？

比如user1，它购买的item顺序是item1,item2,...,item10，那么对于user1，[item1,...,item7]在train，item8在validation，[item9,item10]在test。

train的时候，我们构造了positive sample是根据[item1,...,item6]预测item7，negative sample比如是[item1,...,item6]预测item20。

我们在validation和test的时候，这个用户的输入还是[item1,...,item6]，要预测的下一个商品包括了[item1,..,item10,item20]，因为在训练的时候有positive sample=[item1,...,item6]预测item7，则validation和test的时候，肯定会输出item7为最高分吧，怎么让模型预测输出item8,9,10呢？

这个问题困扰了我好久，如果能帮我解答一下，将万分感谢！

**Comment by graytowne** [comment]
@01joy Hello 感谢您关注我们工作！我不确定我有没有理解您想问什么但是我觉得应该是这样的：
train的时候[item1-item6]预测item7，validation的时候应该是[item1-item7]预测item8

**Comment by 01joy** [comment]
@graytowne 谢谢解答，我大概明白了。
